### PR TITLE
Geometry_Engine: Failure checks and adjustments for FreeFormProfile()

### LINKED ISSUE #1243: Structure_Engine: FreeFormProfile not working in "wrong" orientation
#### Description:
When inputting a profile curve into `FreeFormProfile` it outputs `null` if the curve is in the wrong orientation

This is also a "problem" in `Geometry.Create.IntegrationSlices` but i'm unsure if there is need for that one to handle curves away from the right orientation.

But would suggest either moving the curves to the right orientation in the code before `Geometry.Create.IntegrationSlices` runs or at least give a error message to the user.

#### How to replicate:
Draw a section in the "wrong" plane and send it into `FreeFormProfile`

I'm also 80% sure this used to work

### LINKED ISSUE #1646: Geometry_Engine: StartPoint missing for Ellipse
#### Description:

This has shown to cause crashes in some cases. Should aim to get this in ASAP.

### PR DESCRIPTION
**Pull request by kThorsager**
### Issues addressed by this PR
Closes #1243 
Closes #1646 

Checks for:

- IsClosed
- Planarity
- CoPlanar with XYPlane
- On XYPlane
- Any curve close to zero area
- Curve curve intersection
- SelfIntersection

Fixes (or attemts to fix):

- Projects the curves on to the biggest ones control-points fitted plane
- Rotates the curves to the XYPlane based on the fitted plane
- Translates the curves to the XYPlane

### Test files
https://burohappold.sharepoint.com/:u:/s/BHoM/EQU1FT2YhP1NmU7H1SfPJEMBKsApKeYBGq2jURUPyYTR0g?e=RQ9eDa

### Changelog
- added `Start/EndPoint` for `Ellipse`
- added `IsSelfIntersecting` for `Ellipse`

### CODE REVIEW

**Comment by kThorsager** [comment]
Changed how it handled edges which were non-co-planar with the XY axis after discussion with @IsakNaslundBh ,

Is ready for review again

File: Geometry_Engine/Create/ShapeProfile.cs [hunk 1]
```diff
@@ -324,9 +324,101 @@ public static FabricatedISectionProfile FabricatedISectionProfile(double height,
 
         /***************************************************/
 
+        [Description("Creates a single FreeFormProfile from a collection of ICurves. " +
+                     "Checks if it's a valid profile and attempts to fix; " + 
+                     " - coplanarity with eachother, by projecting onto the plane of the curve with the biggest area. " + 
+                     " - coplanarity with XYPlane, by rotating the curves to align. " + 
+                     " - curves on the XYPlane, by translating from one controlpoint to the origin. " +
+                     "Checks if it's a valid profile; If they're closed, Not zero area, curve curve intersections, selfintersections. ")]
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Create/ShapeProfile.cs:332]:
```suggestion
        [Description("Creates a single FreeFormProfile from a collection of ICurves. \n" +
                     "Checks if it's a valid profile and attempts to fix;  \n" + 
                     " - coplanarity with eachother, by projecting onto the plane of the curve with the biggest area. \n" + 
                     " - coplanarity with XYPlane, by rotating the curves to align. \n" + 
                     " - curves on the XYPlane, by translating from one controlpoint to the origin. \n" +
                     "Checks if it's a valid profile; If they're closed, Not zero area, curve curve intersections, selfintersections. ")]
```

File: Geometry_Engine/Create/ShapeProfile.cs [hunk 2]
```diff
@@ -324,9 +324,101 @@ public static FabricatedISectionProfile FabricatedISectionProfile(double height,
...
+            IEnumerable<ICurve> result = edges.ToList();
+            
+            List<Point> cPoints = edges.SelectMany(x => x.IControlPoints()).ToList();
+
+            // Any Point not on the XY-Plane
+            if (cPoints.Any(x => Math.Abs(x.Z) > Tolerance.Distance))
+            {
+                // Is Planar
+                Plane plane = Compute.FitPlane(cPoints);
+                if (cPoints.Any(x => x.Distance(plane) > Tolerance.Distance))
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Create/ShapeProfile.cs:344]: The content in here is not supported yet for nurbs. If this happens, and we should exit out this larger if statement, as the orientation methods below wont give predictable results if this happens.

File: Geometry_Engine/Create/ShapeProfile.cs [hunk 3]
```diff
@@ -324,9 +324,107 @@ public static FabricatedISectionProfile FabricatedISectionProfile(double height,
...
+                        Reflection.Compute.RecordWarning("The Profiles curves have been projected onto a plane fitted through the biggest curve's control points.");
+                        cPoints = result.SelectMany(x => x.IControlPoints()).ToList();
+                    }
+                    catch
+                    {
+                        failedProject = true;
+                    }
+                }
+
+                if (failedProject)
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Create/ShapeProfile.cs:363]:
```suggestion
                if (!failedProject)
```

**Comment by IsakNaslundBh** [comment]
As discussed Kalle, Raised issue for missing start point method for ellipse here https://github.com/BHoM/BHoM_Engine/issues/1646

Might as well get this fixed in this PR to unblock and merge.

**Comment by IsakNaslundBh** [comment]
I have reviewed the compliance report, and what is flagged is far out of scope of this PR. Will allow this to be merged without any action.

**Comment by IsakNaslundBh** [comment]
Issue regarding compliance breaches raised herehttps://github.com/BHoM/BHoM_Engine/issues/1648

**Pull request merged**